File: 2024/4/soln.cpp
#include<bits/stdc++.h>
using namespace std;

int get_count(vector<vector<char>> v, int i, int j, int m, int n) {
	vector<vector<int>> d = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}, {1, 1}, {-1, -1}, {-1, 1}, {1, -1}};
	int ans = 0;
	for (vector<int> p: d) {
		int r = p[0];
		int c = p[1];
		string next = "XMAS";
		int mul = 0;
		for (mul = 0; mul <= 3; ++mul) {
			if (i + r*mul >= m || j + c*mul >= n || i + r*mul < 0 || j + c*mul < 0) {
				break;
			}
			char chr = v[i + r*mul][j + c*mul];
			if (chr != next[mul]) {
				break;
			}
		}
		if (mul == 4) {
			cout << v[i][j] << "\n";
			ans++;
		}
	}
	return ans;
}

int main() {
	vector<vector<char>> v;
	int i = 0;
	string line;
	ifstream myfile("input.txt");

	if (myfile.is_open()) {
	  while (!myfile.eof()) {
	    getline(myfile, line);
	    // cout << line << endl;
	    v.push_back({});
	    for (char c: line) {
		v[i].push_back(c);
	    }
	    i++;
	  }
	  myfile.close();
	}
	int m = i;
	int n = v[0].size();
	
	int ans = 0;
	for (int i = 0; i < m; ++i) {
		for (int j = 0; j < n; ++j) {
			ans += get_count(v, i, j, m, n);
		}
	}
	cout << ans << "\n";
}

File: 2024/4/soln.zig
const std = @import("std");
const ul = @import("ul");

const xmas = "XMAS";

pub fn main() !void {
    const allocator = std.heap.page_allocator;
    const file = try std.fs.cwd().openFile("input.txt", .{ .mode = .read_only });
    defer file.close();
    var lines = std.ArrayList([]u8).init(allocator);
    defer lines.deinit();
    const buffer = try ul.fileReadLines(allocator, file, &lines);
    defer allocator.free(buffer);

    var ans: u64 = 0;
    for (lines.items, 0..) |line, r| {
        for (line, 0..) |_, c| {
            var dirs = std.ArrayList([2]i8).init(allocator);
            defer dirs.deinit();
            try ul.padj8(&lines, r, c, &dirs);

            for (dirs.items) |dir| {
                var currentR = r;
                var currentC = c;
                var xmasPtr: usize = 0;
                while (xmasPtr != xmas.len and lines.items[currentR][currentC] == xmas[xmasPtr]) {
                    xmasPtr += 1;
                    if (!ul.inBounds(&lines, currentR, currentC, dir[0], dir[1])) {
                        break;
                    }
                    currentR = @as(usize, @intCast(@as(isize, @intCast(currentR)) + dir[0]));
                    currentC = @as(usize, @intCast(@as(isize, @intCast(currentC)) + dir[1]));
                }
                if (xmasPtr == xmas.len) {
                    ans += 1;
                }
            }
        }
    }
    std.debug.print("{d}\n", .{ans});
}
